### PR TITLE
Set onFocusChanged to false when detached

### LINKED ISSUE #156285: Focus: onFocusChanged(false) isn't called when Focus widget is detached
### Steps to reproduce

In this demo https://dartpad.dev/?id=203fe4275aa7980f7ae568c41f5d96eb you'll find that if when you type anything into the input triggering its deletion and the deletion of the Focus widget, the Focus widget's onFocusChange handler isn't called.

### Expected results

There is a reasonable expectation that when focus moves away from a focused Focus, onFocusChange(false) should be called on that node.

### Actual results

focus moves away, but onFocusChange(false) is not called in this situation

(In my usecase, I'm genuinely not sure what to do about this. I tried moving a call to the unfocus code to dispose, but apparently you can't call Providers from dispose)

### Code sample

https://dartpad.dev/?id=203fe4275aa7980f7ae568c41f5d96eb

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.24.1, on NixOS 23.11 (Tapir) 6.7.9, locale en_NZ.UTF-8)
    • Flutter version 3.24.1 on channel stable at
      /nix/store/ivslg1wbqmfcfhaxabdi36bd3cjvhnbq-flutter-wrapped-3.24.1-sdk-links
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision nixpkgs000 (), 1970-01-01 00:00:00
    • Engine revision c9b9d5780d
    • Dart version 3.5.1
    • DevTools version 2.37.2

[✗] Android toolchain - develop for Android devices
    • Android SDK at /home/mako/Android/Sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.

[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Linux toolchain - develop for Linux desktop
    • clang version 18.1.8
    • cmake version 3.27.7
    • ninja version 1.12.1
    • pkg-config version 0.29.2

[!] Android Studio (version 2022.2)
    • Android Studio at
      /nix/store/nj9ncfr3mgwly9gagj6m59if5gp37k1g-android-studio-stable-2022.2.1.19-unwrapped
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Failed to run Java: ProcessException: No such file or directory
        Command:
        /nix/store/nj9ncfr3mgwly9gagj6m59if5gp37k1g-android-studio-stable-2022.2.1.19-unwrapped/jbr/bin/ja
        va -version
    ✗ Unable to determine bundled Java version.
    • Try updating or re-installing Android Studio.

[☠] Connected device (the doctor check crashed)
    ✗ Due to an error, the doctor check did not complete. If the error message below is not helpful,
      please let us know about this issue at https://github.com/flutter/flutter/issues.
    ✗ Error: Unable to run "adb", check your Android SDK installation and ANDROID_HOME environment
      variable: /home/mako/Android/Sdk/platform-tools/adb
      Error details: No such file or directory
    • #0      throwToolExit (package:flutter_tools/src/base/common.dart:10:3)
      #1      AndroidDevices.pollingGetDevices
      (package:flutter_tools/src/android/android_device_discovery.dart:75:7)
      <asynchronous suspension>
      #2      PollingDeviceDiscovery._populateDevices (package:flutter_tools/src/device.dart:548:36)
      <asynchronous suspension>
      #3      Future.wait.<anonymous closure> (dart:async/future.dart:534:21)
      <asynchronous suspension>
      #4      DeviceManager.refreshAllDevices (package:flutter_tools/src/device.dart:218:40)
      <asynchronous suspension>
      #5      DeviceValidator.validate (package:flutter_tools/src/doctor.dart:714:34)
      <asynchronous suspension>
      #6      Future.any.onValue (dart:async/future.dart:628:5)
      <asynchronous suspension>


[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 4 categories.

```

</details>

### PR DESCRIPTION
**Pull request by RamonFarizel**
According to the issue description we should call `onFocusChanged(false)` before detaching the widget.
This is my first attempt to solve this issue, but I not sure about its reliability. 
Since I was not sure about where ask for help I decided to push this first approach and fix after receiving guidance.

Thanks in advance! 

Fixes #156285

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by nate-thegrate — COMMENTED** [review]
Thanks for making this contribution!

Unfortunately, I'm kind of at a loss here—not sure how we could solve this issue without breaking things 🙁

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 1]
```diff
@@ -566,6 +566,7 @@ class _FocusState extends State<Focus> {
   void dispose() {
     // Regardless of the node owner, we need to remove it from the tree and stop
     // listening to it.
+    widget.onFocusChange?.call(false);
```
[the remark this answers is not in the document]
Comment (RamonFarizel) [reply packages/flutter/lib/src/widgets/focus_scope.dart:569]: I see. 

> Unfortunately, I'm kind of at a loss here—not sure how we could solve this issue without breaking things 🙁
Do you think we should close this PR? If so what could be done with the issue? 

Just sharing some thoughts about the line below.
`focusNode.removeListener(_handleFocusChanged);`

it sounds like when we trigger the `_handleFocusChanged` the listener was already clean up. So the callback inside it won't work.
And even if it worked the focusNode is being removed so its value is not being updated.

It makes me wonder if it is a bug or just the regular behavior.

[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/widgets/focus_scope.dart:569]: @gspencergoog WDYT? Does this sound like the right behavior?

[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/focus_scope.dart:569]: You don't need to call this explicitly. If you want to have that happen, just move the next line (the `removeListener` line) down one to below the `detach`, since detaching the attachment will force the node to give up focus and if the listener is still registered, it'll call it. You'd also be bypassing some code that runs in `_handleFocusChanged` if you call `widget.onFocusChange` directly.

However, I think that will cause a bunch of handlers to get called during dispose when they aren't expecting to, and they'll try to call `setState` and throw.

You could try and clean up those instances in the framework and see how that works for you, but it feels like something that will break a lot of code out there.

[the remark this answers is not in the document]
Comment (RamonFarizel) [reply packages/flutter/lib/src/widgets/focus_scope.dart]: Thanks for your help! 
I moved the `removeListener` below the `detach`, but it seems like I'm still getting the same result. When the widget is removed, the focus change doesn't trigger. 
By the time we reach the line `widget.onFocusChange?.call(focusNode.hasFocus);`, `widget.onFocusChange` is already null.

[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/focus_scope.dart]: Ahh, yeah, actually that makes sense.  The unfocus isn't going to actually take effect until after the current frame is done, so by that time the widget has already been disposed.

Hmm. Well, I guess you could try and call `widget.onFocusChange` directly here, but I still think that will cause problems. No harm in trying it, though. I would only call it if the node currently has focus.

[the remark this answers is not in the document]
Comment (RamonFarizel) [reply packages/flutter/lib/src/widgets/focus_scope.dart]: There's still something odd happening. The dispose method is being triggered twice. Is this expected behavior?

During the first call, focusNode.hasPrimaryFocus is false, and widget.onFocusChange is null.
During the second call, focusNode.hasPrimaryFocus is false, and widget.onFocusChange is `not` null.

if the following conditional I was able to make it work:
```
if (!focusNode.hasFocus && widget.onFocusChange != null) {
      widget.onFocusChange?.call(false);
    }
```

I didn't see any exceptions for now.

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 2]
```diff
@@ -564,10 +564,14 @@ class _FocusState extends State<Focus> {
...
   void dispose() {
+    if (!focusNode.hasFocus && widget.onFocusChange != null) {
+      widget.onFocusChange?.call(false);
+    }
+
     // Regardless of the node owner, we need to remove it from the tree and stop
     // listening to it.
-    focusNode.removeListener(_handleFocusChanged);
     _focusAttachment!.detach();
+    focusNode.removeListener(_handleFocusChanged);
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/widgets/focus_scope.dart:574]: Thanks for updating the setup! Unfortunately it looks like many checks are still failing:

some checks were not successful

<br>

Triggering the `onFocusChange` callback as the Focus is being disposed of could be a massively breaking change, since I bet there are many existing callbacks inside & outside the framework that involve `setState` calls.

My preference would be to close this PR and recommend that a developer supply a focus node if this sort of behavior is desired.

Comment (RamonFarizel) [reply packages/flutter/lib/src/widgets/focus_scope.dart:574]: I agree, let's close this PR for now. It was a productive discussion!

How should we respond in the issue section? Should I leave a comment to explain the decision?

Comment (nate-thegrate) [reply packages/flutter/lib/src/widgets/focus_scope.dart:574]: > I agree, let's close this PR for now. It was a productive discussion!

Love this energy!

<br>

> How should we respond in the issue section?

I already went ahead and left a comment there: https://github.com/flutter/flutter/issues/156285#issuecomment-2417956193
We can probably just wait and see how they choose to respond :)

**Comment by nate-thegrate** [comment]
Closing per https://github.com/flutter/flutter/pull/156779#discussion_r1803602450.
